JAVA-846: Provide custom codecs library as an extra module

adutra opened the pull request:
Also includes the following tickets:
- JAVA-742 (JSON codecs)
- JAVA-606 (Java 8 codecs)
- JAVA-565 (Array codecs)
- JAVA-605 (Enum codecs and mapper support)

Commit:
JAVA-846: Provide custom codecs library as an extra module.

This commit contains substantial contributions by Olivier Michallat (@olim7t <omichallat+github@gmail.com>) and Andrew Tolbert (@tolbertam <andrew.tolbert@datastax.com>).

This commit creates a new 'driver-extras' Maven module that will host goodies for the driver,
such as extra codecs. Note that Java >= 8 is required to compile this module.
It also updates Maven dependencies to more recent versions.

It also updates the Maven clirr plugin configuration with a workaround that
makes it compatible with Java 8.
The bug is actually in the BCEL library, see https://issues.apache.org/jira/browse/BCEL-173.
The workaround itself has been suggested here: https://github.com/RichardWarburton/lambda-behave/issues/31#issuecomment-86052095.

It also adds Maven animal-sniffer plugin to verify that compiled classes remain compatible with the driver's target API (Java 6).
Execution of the plugin is disabled for the new driver-extras module has this module currently requires Java 8, but there is no signature available yet for this JDK.

This commit introduces the following new features:

- AbstractCollectionCodec, AbstractMapCodec, AbstractUDTCodec and AbstractTupleCodec.
- MappingCodec, a new general-purpose codec that maps Pojos to another intermediary Java object that is in turn mapped to a CQL type.
- ParsingCodec, a new general-purpose codec that maps Pojos into Strings that can in turn be mapped to a CQL varchar type.
- Two codecs that support automatic conversion from Java enums to CQL varchar or int types.
- Codec support for Guava's Optional API.
- Codec support for Joda Time. This commit introduces new codecs to automatically map Joda Time types such
as LocalTime, LocalDate, Instant and DateTime to CQL types (respectively, time, date,
timestamp and tuple<timestamp,varchar>).
- SimpleTimestampCodec: maps from CQL timestamp to Java long (representing milliseconds since the Epoch);
- SimpleDateCodec: maps from CQL date to Java int (representing days since the Epoch).

Remarks:
- This commit modifies ParseUtils and CodecUtils and makes them public.
- This commit also comes with some code cleanup in TypeCodec.
- This commit also fixes wrong imports in MapperEnumTest. The file was importing jUnit's Assert class instead of TestNG's.

Commit:
JAVA-742: Codec support for JSON.

This commit introduces two codecs:

- a new JacksonJsonCodec that automatically maps JSON structures to CQL varchar types leveraging the popular Jackson framework.
- a new Jsr353JsonCodec that automatically maps JSON structures to CQL varchar types using JSR 353.

Commit:
JAVA-606: Codec support for Java 8.

This commit introduces new codecs to automatically map Java 8 Date & Time API types such
as LocalTime, LocalDate, Instant and ZonedDateTime to CQL types (respectively, time, date,
timestamp and tuple<timestamp,varchar>).

It also introduces a new codec to support Java 8 Optional API.

Commit:
JAVA-565: Codec support for Java arrays.

This commit introduces codecs that map CQL types to Java arrays directly. Object arrays and arrays of primitive types are both supported.

olim7t commented on driver-extras/src/main/java/com/datastax/driver/extras/codecs/ArrayCodecs.java:
```diff
@@ -0,0 +1,583 @@
+ * be registered via {@link #withPrimitiveArrayCodecs(CodecRegistry)}.
+ *
+ */
+public abstract class ArrayCodecs {
```
We could use sub-packages instead of nesting stuff into abstract classes.

adutra replied:
Very good idea, done.

olim7t commented on driver-extras/src/main/java/com/datastax/driver/extras/codecs/ArrayCodecs.java:
```diff
@@ -0,0 +1,583 @@
+     */
+    public static CodecRegistry withPrimitiveArrayCodecs(CodecRegistry registry) {
+        return registry.register(IntArrayCodec.instance, LongArrayCodec.instance, FloatArrayCodec.instance, DoubleArrayCodec.instance);
+    }
```
I'm not a big fan of this, because the client-side syntax is inconsistent:

``` java
registry.register(new MyCustomCodec());
ArrayCodecs.withPrimitiveArrayCodecs(registry);
```

I'd rather have a method that returns an iterable, that we can pass to register:

``` java
registry.register(new MyCustomCodec());
registry.register(ArrayCodecs.all());
```

adutra replied:
After splitting the classes into separate files, it seemed to me that it is not necessary to have such shortcuts.

tolbertam commented on .travis.yml:
```diff
@@ -1,5 +1,8 @@
 language: java
 jdk:
   - openjdk6
+    script: mvn test -DlegacyJdk=true
```
I think you will also want to do:

``` yaml
install: true
```

to bypass mvn install (which mvn test will do anyways).

Commit:
JAVA-605: Codec support for Java enums.

This commit drops the mapper @Enumerated annotation. Users are now encouraged to use the enum codecs instead (EnumNameCodec and EnumOrdinalCodec).

olim7t commented:
`CodecUtils` has two distinct responsibilities: ser/deser stuff, and type token helpers. How about we extract the token methods in a separate class, for instance `TypeTokens`?

olim7t committed:
JAVA-884: Fix UDT mapper to process fields in the correct order.

olim7t commented:
Pushed a new commit to fix JAVA-884 (also makes `MapperUDTCodec` extend `AbstractUDTCodec`). Note that this requires JAVA-936 to work with Cassandra 3.0.0 GA, otherwise it should be tested against rc2.

tolbertam commented on driver-extras/src/main/java/com/datastax/driver/extras/codecs/jdk8/InstantCodec.java:
```diff
@@ -0,0 +1,134 @@
+        if (value == null || value.isEmpty() || value.equalsIgnoreCase("NULL"))
+            return null;
+        // strip enclosing single quotes, if any
+        if (value.charAt(0) == '\'' && value.charAt(value.length() - 1) == '\'')
```
ParseUtils.isQuoted would be good for this.

tolbertam replied:
Ah nevermind, these are single quotes.

adutra replied:
Good catch thanks!

tolbertam commented on driver-extras/src/main/java/com/datastax/driver/extras/codecs/jdk8/LocalDateCodec.java:
```diff
@@ -0,0 +1,103 @@
+    public String format(LocalDate value) {
+        if (value == null)
+            return "NULL";
+        return "'" + ISO_LOCAL_DATE.format(value) + "'";
```
PraseUtils.quote would be good for this

adutra replied:
same here :)
